README.txt
=Features=
   Support binary GPIOs, in both input and output.
   HTML / Javascript / CSS client side
   Easily [CUSTOMIZE] UI
   REST/JSON [WEBAPI]
   Server side available in two technologies
     PHP 
     Python
   Smartphone compatible
   Auto-refresh

=Restrictions=
   GPIOs 0 and 1 have input pull-up resistor enabled.
   UART is enabled but unusable for now.
     GPIOs 14 and 15 are disabled
   UART, SPI and I2C support will be added as soon as possible.
   Don't forget that GPIO support only 3.3V, don't plug them to 5V !!!

=Usage=
First of all, follow [INSTALL] instructions.

If your are directly using your Raspberry Pi with keyboard/mouse/display plugged,  open a browser to http://localhost/webiopi/

Don't forget to add :port right after raspberrypi/IP if you use a different port.

If your Raspberry Pi is connected to your network, yo can open a browser to http://raspberrypi/webiopi/ with any device of your network. 
Replace raspberrypi by its IP if DNS don't find raspberrypi.

You can even add a port redirection on your router (and/or use IPv6) to control your GPIOs over Internet !

   Click/Tap the OUT/IN button to change GPIO direction.
   Click/Tap pins to change the GPIO output state.

http://trouch.com/wp-content/uploads/2012/08/webiopi-chrome.png

------------------------------------------------------------------------------

   Copyright 2012 Eric Ptak - trouch.com

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.



doc/CHANGELOG
==0.6 (27/03/13)==
   Added CoAP ([http://tools.ietf.org/html/draft-ietf-core-coap-14 draft-14]) implementation (Server and Client)
   Added Python WebIOPi Client class (HTTP or CoAP with HTTP fallback)
   Added Python WebIOPi MulticastClient class (CoAP)
   Added Serial, I2C, SPI and 1-Wire  support
     Python lightweight drivers with no dependency
     Automatically load required linux modules
     Added many device drivers, see [DEVICES device support page]
   Added [Configuration] file
   Added Serial monitor web app
   Added Devices monitor web app
   Added Devices abstractions Javascript class
   Added GPIO digitalRead/Write and getFunction methods
   Added Cache-Control REST response header
   Added Server logging facility
   Added jQuery-mobile
   Improved button up/down handling on mobile devices
   Fixed error fetching python-dev package during setup
   Fixed webiopi-passwd missing execution flag when installing with PiStore
   Fixed binary file serving
   Fixed iOS6 Safari Mobile bug


==0.5.3 (01/07/13)==
   Added board revision REST URI (submitted by Andreas Riegg)
   Added encrypted passwd file to store credentials used for HTTP authentication
   Added webiopi-passwd command-line program to generate passwd file
   Moved demo to examples/custom folder, added examples/basic
   Changed /dev/mem access to allow webiopi import without root privileges when not using GPIOs
   Disabled update checker to avoid "Update available" link
   Improved GPIO error handling
   Fixed encoding issue with python 2.x giving a blank page
   Fixed setup script

==0.5.2 (12/21/12 - PiStore only)==
   Fixed blank page and file handling when server start at boot
   Improved macros handling to allow zeo, one or more args
   Added server loop helper
   Improved setup.sh for the PiStore
   Added play.sh for the PiStore to open the browser

==0.5.1 (11/16/12)==
===New Features===
   REV 2 boards support
   Added setup script to ease WebIOPi install
   Use WebIOPi in your own Python scripts
   Login/Password protection
   Software PWM
   Binary sequence output

===Python Server & REST API===
   Usable as a library
   Added Python 3 support
   Removed RPi.GPIO library dependency
   Improved security
   Improved file serving
   Improved REST API
   Added ability to use custom REST macro
   Added ability to output a single pulse
   Added ability to output a binary sequence
   Added software PWM

===Javascript Library===
   Improved and simplified
   Added ability to create custom buttons with one or two callbacks (mousedown and mouseup)
   Added helpers for new REST functions (macro, pulse, sequence, PWM, ...)

===Other changes===  
   PHP Server discontinued


==0.3 (08/27/12)==
   Fixed security issue in the Python server
   Python server general improvements
   Python server display real IP
   Allow PHP5 < 5.4 (Works on Debian wheezy)
   Added update tracking
   Refactored Javascript code to be used as a library
   Changed IDs and CSS naming
   Added expert app


==0.2 (08/21/12)==
   Initial release

gpio.php
<?php
/*
   Copyright 2012 Eric Ptak - trouch.com

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

class GPIO {
	var $GPIO_AVAILABLE = array(0, 1, 4, 7, 8, 9, 10, 11, 14, 15, 17, 18, 21, 22, 23, 24, 25);
	var $ALT = array(
		"UART" => array("enabled" => true, "pins"=> array(14, 15)),
		"I2C" => array("enabled" => false, "pins"=> array(0, 1)),
		"SPI" => array("enabled" => false, "pins"=> array(7, 8, 9, 10, 11))
	);
	var $GPIO_ROOT = "/sys/class/gpio";
	
	function sudo($cmd) {
		return system("sudo sh -c '$cmd'");
	}
	
	function setALT($alt, $enable) {
		$alt["enabled"] = $enable;
	}

	function getPath($pin, $cmd="") {
		return $this->GPIO_ROOT . "/gpio" . $pin . "/" . $cmd;
	}

	function isEnabled($pin) {
		return file_exists($this->getPath($pin));
	}

	function enable($pin) {
		$this->sudo("echo $pin > $this->GPIO_ROOT/export");
		if ($this->isEnabled($pin)) {
			$this->setDirection($pin, "in");
			return true;
		}
		return false;
	}

	function disable($pin) {
		$this->sudo("echo $pin > $this->GPIO_ROOT/unexport");
		return $this->isEnabled($pin);
	}

	function readGPIO($pin, $file) {
		$path = $this->getPath($pin, $file);
		$f= fopen($path, "r");
		$value = fread($f, filesize($path));
		$value = substr($value, 0, strpos($value, "\n"));
		fclose($f);
		return $value;
		
	}
	
	function writeGPIO($pin, $file, $value) {
		$this->sudo("echo $value > " . $this->getPath($pin, $file));
		return $this->readGPIO($pin, $file);
	}
	
	function getDirection($pin) {
		return $this->readGPIO($pin, "direction");
	}

	function setDirection($pin, $value) {
		return $this->writeGPIO($pin, "direction", $value);
	}

	function getValue($pin) {
		return $this->readGPIO($pin, "value");
	}

	function setValue($pin, $value) {
		return $this->writeGPIO($pin, "value", $value);
	}

	function getALTPins() {
		$alt_pins = array();
		foreach($this->ALT as $alt) {
			if ($alt["enabled"]) {
				foreach($alt["pins"] as $pin) {
					$alt_pins[] = $pin;
				}
			}
		}
		return $alt_pins;
	}

	function setPINSEnabled($enable) {
		$alt_pins = $this->getALTPins();
		foreach($this->GPIO_AVAILABLE as $pin) {
			if (!in_array($pin, $alt_pins)) {
				if ($enable && !$this->isEnabled($pin)) {
					$this->enable($pin);
				}
				else if (!$enable && $this->isEnabled($pin)) {
					$this->disable($pin);
				}
			}
		}

	}

	function init() {
		$this->setPINSEnabled(true);
	}

	function release() {
		$this->setPINSEnabled(false);
	}
}

if (isset($_SERVER["SHELL"]) && !isset($_SERVER["GATEWAY_INTERFACE"])) {
	$gpio = new GPIO();
	if ($argc == 1) {
		
	}
	else if (($argc == 2) && ($argv[1] == "init")) {
		$gpio->init();
	}
	else if (($argc == 2) && ($argv[1] == "release")) {
		$gpio->release();
	}
	else {
		echo "usage:\n";
		echo "php " . $argv['0'] . " [init|release]\n";
	}
}
?>
